### ISSUE TITLE
[MSVC2013]'Catch::AutoReg::AutoReg' : no overloaded function takes 5 arguments

### ISSUE THREAD

**Issue by RossBencina** [philsquared/Catch#549]
On develop branch at current commit: d274fc5

When compling the tests using MSVS2013 Update 4, using a solution generated by CMake, I get the following error:

>C:\catch\Catch-develop-appveyor\projects\SelfTest\TestMain.cpp(458): error C2661: 'Catch::AutoReg::AutoReg' : no overloaded function takes 5 arguments

Workaround: #ifdef out ManuallyRegistered test case in TestMain.cpp

**Comment by philsquared** [comment]
I forgot to follow up - are you still seeing this issue?

**Comment by RossBencina** [comment]
No, it still fails with master 1.3.1.

Steps to reproduce are:

 * Run the following from the command line.

> git clone https://github.com/philsquared/Catch.git
> cd Catch\projects\CMake
> cmake -G"Visual Studio 12 2013"
> start Catch.sln

 * Once MSVS has launched, click the run button.

Results in the same error as above:

> 1>------ Build started: Project: ZERO_CHECK, Configuration: Debug Win32 ------
1>Build started 31/12/2015 3:14:05 PM.
1>PrepareForBuild:
1>  Creating directory "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\".
1>InitializeBuildStatus:
1>  Creating "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
1>CustomBuild:
1>  Checking Build System
1>  CMake does not need to re-run because C:/catch/Catch/projects/CMake/CMakeFiles/generate.stamp is up-to-date.
1>FinalizeBuildStatus:
1>  Deleting file "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
1>  Touching "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
1>
1>Build succeeded.
1>
1>Time Elapsed 00:00:02.29
2>------ Build started: Project: SelfTest, Configuration: Debug Win32 ------
2>Build started 31/12/2015 3:14:08 PM.
2>InitializeBuildStatus:
2>  Creating "SelfTest.dir\Debug\SelfTest.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
2>CustomBuild:
2>  Building Custom Rule C:/catch/Catch/projects/CMake/CMakeLists.txt
2>  CMake does not need to re-run because C:\catch\Catch\projects\CMake\CMakeFiles\generate.stamp is up-to-date.
2>ClCompile:
2>  ApproxTests.cpp
2>  BDDTests.cpp
2>  ClassTests.cpp
2>  ConditionTests.cpp
2>  ExceptionTests.cpp
2>  GeneratorTests.cpp
2>  MessageTests.cpp
2>  MiscTests.cpp
2>  PartTrackerTests.cpp
2>  TestMain.cpp
2>C:\catch\Catch\projects\SelfTest\TestMain.cpp(458): error C2661: 'Catch::AutoReg::AutoReg' : no overloaded function takes 5 arguments
2>  TrickyTests.cpp
2>  VariadicMacrosTests.cpp
2>  EnumToString.cpp
2>  ToStringPair.cpp
2>  ToStringVector.cpp
2>  ToStringWhich.cpp
2>  ToStringTuple.cpp
2>  Generating Code...
2>
2>Build FAILED.
2>
2>Time Elapsed 00:00:22.76
3>------ Skipped Build: Project: ALL_BUILD, Configuration: Debug Win32 ------
3>Project not selected to build for this solution configuration

**Comment by RossBencina** [comment]
Intellisense error gives more details:

> IntelliSense: no instance of constructor "Catch::AutoReg::AutoReg" matches the argument list
            argument types are: (void (), const char [19], const char [1], Catch::SourceLineInfo, Catch::NameAndDesc)	c:\catch\Catch\projects\SelfTest\TestMain.cpp	458

The problem seems to be with the `CATCH_CONFIG_VARIADIC_MACROS` version of `INTERNAL_CATCH_REGISTER_TESTCASE` in catch_test_registry.hpp:

```
#define INTERNAL_CATCH_REGISTER_TESTCASE( Function, ... ) \
       Catch::AutoReg( Function, CATCH_INTERNAL_LINEINFO, Catch::NameAndDesc( __VA_ARGS__ ) );
```

It looks like the `Function` parameter to the macro `INTERNAL_CATCH_REGISTER_TESTCASE` is getting all three of the parameters (i.e. (void (), const char [19], const char [1]) not just the first one as intended.

Maybe you're using a non-standard variadic macros form?

**Comment by RossBencina** [comment]
It seems to boil down to the following failing in MSVC:

```
#include <cstdio>
 
void internal_foo(int x)
{
    std::printf("%d\n", x);
}
 
#define INTERNAL_CALL_FOO(X, ...) internal_foo(X)
#define CALL_FOO(...) INTERNAL_CALL_FOO(__VA_ARGS__ )
int main(int argc, char* const argv[])
{
    CALL_FOO(42, 0, 0);
}
```

Fails compilation with:

```
 		2	IntelliSense: too many arguments in function call	c:\testmain.cpp	12
Error	1	error C2660: 'internal_foo' : function does not take 3 arguments	c:\testmain.cpp	12
```

The same code works fine with clang and gcc on godbolt and ideone.

**Comment by RossBencina** [comment]
It turns out that the desired behavior is not supported by MSVC. This was previously documented in (now deleted) Microsoft Connect issue 380090 [1]. It has been discussed in the context of many projects:

https://www.google.com/search?q=380090+__va_args__

There is a discussion of workarounds here:
http://stackoverflow.com/questions/9183993/msvc-variadic-macro-expansion

I have not read through the workarounds to evaluate how complex the fix would be.

How important is use of `__VA_ARGS__`  here? With Catch, perhaps the best workaround may be to disable use of this idiom with MSVC?

[1] Thanks to glenfe_ on freenode ##c++ for this info.

**Comment by RossBencina** [comment]
A simple fix might be to change the definition of `REGISTER_TEST_CASE` to the following, which has the same form as the other registration macros:

```
#define REGISTER_TEST_CASE( function, ... ) INTERNAL_CATCH_REGISTER_TESTCASE( function, __VA_ARGS__ )
```

This compiles fine with MSVC 2013.

**Comment by philsquared** [comment]
OIC. That makes sense (in a buggy, VS2013, kind of way).
I hadn't noticed that I'd written `REGISTER_TEST_CASE` differently to the other forwarding macros - which explains why I hadn't seen this issue before (I use Catch with VS2013 daily - but have never used `REGISTER_TEST_CASE` outside of the test project).
I've committed a fix, along the lines of your suggestion - which brings it into line with the other forwarding macros. I've only tested that with clang so far but it sounds like you've already verified the equivalent with VS2013.
I've not done a single header release with it in yet - but if you're able to verify it in the context of the multiple-headers test project that would be great - thanks!
(I also only remembered, after pushing, that I hadn't fixed `CATCH_REGISTER_TEST_CASE` so I've done that now too, locally).
